## 7. 오류 처리

### 오류 코드 보다는 예외를 사용해라
아래와 같은 방법을 사용하면 함수를 호출한 즉시 오류를 확인해야 하기 때문에 호출자 코드가 복잡해진다.

1. 호출한 즉시 오류를 확인해야 하기 때문에 *호출자 코드가 복잡* 해진다.
2. 오류 확인을 잊어버리기 쉽다.

```swift
func sendShutDown() {
  let handle: DeviceHandle = getHandle(DEV1)
  if handle != DeviceHandle.INVALID {
    retrieveDeviceRecord(handle)
    
    if record != DEVICE_SUSPENDED {
      pauseDevice(handle)
      clearDeviceWorkQueue(handle)
    } else {
      logger.log("Devicesu spended. Unable to shut down")
    }

  } else {
    logger.log("Invalid handle for: " + DEV1.toString())
  }
}
```

아래와 같이 논리가 오류 처리 코드와 섞이지 않아서 깔끔해지기 때문에 오류가 발생하는 코드는 예외를 던지는 편이 좋다.

예외가 발생할 코드를 짤때는 do-catch(java는 try-catch-finally) 문으로 시작하는 것이 좋다.

```swift
public func sendShutDown() throws {
  do {
    try tryToShutDown()
  } catch error {
    logger.log(error)
  }
}

private func tryToShutDown() throws {
  DeviceHandle handle = getHandle(DEV1);
  DeviceRecord record = retrieveDeviceRecord(handle);
  
  pauseDevice(handle);
  clearDeviceWorkQueue(handle);
  closeDevice(handle);
}

private func getHandle(DeviceId id) throws -> DeviceHandle {
  //....
  throw DeviceShutDownError
}
```

### **Unchecked(미확인) 예외를 사용하라**
- 잠깐! 확인, 미확인 예외가 뭐야  
    * 예외는 2가지로 구분됨
    * 확인된 예외(Checked Excaption), 미확인 예외(Unchecked Exception)
    
    <br>

    - **확인된 예외(Checked Exception)**
        - 잘못된 코드가 아닌 잘못된 상황에서 발생하는 예외
        - 예를 들어 파일 열기와 같은 정확한 코드로 구현 했음에도, 외부환경(파일이 없는것)에 따라 발생 가능
        - RuntimeException이외의 예외들
    - **미확인 예외(Unchecked Exception)**
        - 컴파일 에러가 나지 않지만 적절한 예외처리가 없을 경우 강제종료등
        - RuntimeException에 포함된 예외들
        - 
**미확인 예외(unchecked exception)** 은 잘못구현된 프로그램 코드 때문에 발생한다. Runtime도중에 발생하는 예외라고 판단할수 있고 예를들어 배열에서 존재하지 않는 인덱스에 접근할때와 같이 발생하는 오류가 예외와 같은 형태라고 볼 수 있다.

**확인된 예외(checked exception)** 은 RuntimeException과 관련 없는 객체들로서 예를들어 존재하지 않는 외부 파일을 사용할 때 발생하는 예외다. 확인된 예외는 컴파일 될때 컴파일러가 예외처리를 했는지 검사하기 때문에 오류 발생할 것을 대비해 예외 처리를 해주어야 한다.

확인된 예외는 OCP(Open Closed Princlple)를 위반하고 캡슐화가 깨진다.
- 예외를 던지는 메서드가 catch블록이 있는 메서드가 아닌 더 하위에 있다면 그 사이의 메서드에서 모두 해당 예외를 선언부에 추가하거나 catch블록에서 처리해야 하야 하기 때문에

### **예외에 의미를 제공하라(예외 메세지)**
예외를 던질 때는 전후 상황을 충분히 덧붙인다.  
연산 이름, 실패 유형 등

### 호출자를 고려해 예외 클래스를 정의하라

오류를 분류하는 방법은 수없이 많다. 예를들어 디바이스 실패, 네티워크 실패 등  
하지만 오류를 잡아내는 방법이야 말로 프로그래머에게 가장 중요하다  
다음은 외부 라이브러리를 사용할 때 Wrapping을 해주면 좋다.

- 에러 처리가 간결해짐
- 외부 라이브러리와 프로그램 사이의 의존성이 크게 줄어듦
- 프로그램 테스트가 쉬워짐
- 외부 API 설계 방식에 의존하지 않아도 됨

**as-is**

```java
ACMPort port = new ACMPort(12);
try{
	port.open();
} catch (DeviceREsponseException e){
	reportPortError(e)
	logger.log("Device response exception", e)
} catch (ATM1212UnlockedException e){
	reportPortError(e)
	logger.log("Unlock Exception", e)
} catch (GMXError e){
	reportPortError(e)
}

```

**to-be**

```java
LocalPort port = new LocalPort(12);

try{
	port.open();
} catch (PortDeviceFailure e){
	reportPortError(e)
	logger.log("Unlock Exception", e)
}

class LocalPort{
	private ACMPort innerPort;
	public LocalPort(int portNumber){
		innerPort = ACMPort(portNumber);	
	}

	public void open(){
		try{
			innerPort.open();
		} catch (DeviceResponseException e){
			throw new PortDeviceFailure(e);
		} catch (ATM1212UnlockedException e){
			throw new PortDeviceFailure(e);
		} catch (GMXError e){
			throw new PortDeviceFailure(e);
		}
	}
}

```

### **null을 반환하지 마라**
null을 반환하는 코드는 일거리를 늘릴 뿐 아니라 호출자에게 문제를 떠넘긴다.
- 호출자에게 null을 체크할 의무를 줌
- *NullPointerException* 의 발생 위험이 있음
- null확인이 너무 많아짐

```java
public void registerItem(Item item) {
	if (item != null) {
		ItemRegistry registry = peristentStore.getItemRegistry();
		if (registry != null) {
			Item existing = registry.getItem(item.getID());
			if (existing.getBillingPeriod().hasRetailOwner()) {
				existing.register(item);
			}
		}
	}
}
```

### **null을 전달하지 마라**
함수의 인수로 null을 넣지 말자.  
함수에서 NullPointException이 발생할 수 있다.  
함수에서 Null Check 후 다른 Exception을 발생시키는 방법이 있다.